src/rust/grapl-observability/src/statsd_formatter.rs

```diff
@@ -0,0 +1,161 @@
+use crate::metric_error::MetricError;
+use failure::Error;
+use itertools::join;
+use lazy_static::lazy_static;
+use regex::Regex;
+
+lazy_static! {
+    static ref INVALID_CHARS: Regex = Regex::new("[|#,=:]").unwrap();
+}
+
+// TODO look at lifetimes, should this own or not, etc
+// seems odd to .to_string() everytihng going in here
+pub struct TagPair {
+    tag_key: String,
+    tag_value: String,
+}
+
+impl TagPair {
+    fn statsd_serialized(&self) -> String {
+        return format!("{}={}", self.tag_key, self.tag_value);
+    }
+}
+
+fn reject_invalid_chars(s: &str) -> Result<(), Error> {
+    let matched = INVALID_CHARS.is_match(s);
+    if matched {
+        Err(Error::from(MetricError {}))
+    } else {
+        Ok(())
+    }
+}
+
+pub enum MetricType {
+    Gauge,
+    Counter,
+    Millis,
+    Histogram,
+}
+
+impl MetricType {
+    fn statsd_type(&self) -> String {
+        match self {
+            MetricType::Gauge => "g",
+            MetricType::Counter => "c",
+            MetricType::Millis => "ms",
+            MetricType::Histogram => "h",
+        }
+        .to_string()
+    }
+}
+
+pub fn statsd_format(
+    metric_name: &str,
+    value: f64, // should i make a union type?
+    typ: MetricType,
+    sample_rate: Option<f64>,
+    tags: &[TagPair],
+) -> Result<String, Error> {
+    // unlikely we'd go over this
+    reject_invalid_chars(metric_name)?;
+
+    let mut sections: Vec<String> = Vec::with_capacity(6);
```

Comment:
If the end goal is to build up a String it's probably simpler to just do String::with_capacity() and use .push_str

It'll allocate less, though it does mean it might be slightly less pretty since you wouldn't do the join("|") but instead just push a "|" where appropriate.